### ISSUE TITLE
Consider switching to pipx commands instead of 'Breeze2` bootstrap

### ISSUE THREAD

**Issue by potiuk** [apache/airflow#20921]
I wanted to discuss an important change proposal for our Breeze2 approach.

TL;DR; I thought a lot about how we shold run Breeze2. Instead of having a single `./breeze` command we could have multiple `airflow-*` commands that would be automatically installed by `pipx install -e ./dev/breeze` and that it should be the only way of running various commands of breeze (both in CI and during local development). An example of that is `airflow-freespace` command that is used in CI to free space needed to run the commands.

# Context

I looked at some of the changes made recently, particularly just merged: https://github.com/apache/airflow/pull/20701 and I thought we might change the approach of the new Breeze following the learnings we had and mostly suggestions of @uranusjr  when we dicussed the custom code in https://github.com/apache/airflow/blob/main/Breeze2 for bootstraping the environment.

This is - in essence very similar aas the old `./breeze` was and I kept to be "anchored" with it however by introducing `pipx` and installing ./breeze locally in editable mode, we've almost achieved what I wanted from the beginning - i.e. very simple and os-independent way of bootstrapping the environment for Breeze.

Current state, reasoning  and decision is captured here (still Draft though): https://github.com/apache/airflow/blob/main/dev/breeze/doc/adr/0003-bootstraping-virtual-environment.md

# Proposal

The "free-space" script shows how we can utilize multiple entry-points,. Instead of "one-command to rule them all" - "Breeze"  - we could have multiple commands/entry-points installed with `pipx`:

* airflow-breeze -> to run Breeze shell
* airflow-build-ci-image -> to build ci image
* airflow-build-prod-image-> to build prod image
* airflow-build-docs - > to build docs,
* airflow-run-static-checks - > to run static checks 

and so on, rather than:

* ./Breeze2 build-ci-image
* ./Breeze2 build-docs

and so on.

# Why it might be better ?

Mainly because of `zen of python`: https://www.python.org/dev/peps/pep-0020/ 

> There should be one-- and preferably only one --obvious way to do it.

Wile `pipx-only` approach is not perfect, I think we can make it close to it (with some effort and added complexity). 

With PIPX installation, all those commands are easily accessible in the PATH (and auto-complete'able).

Also the very same commands could be used in CI (same as we did with "airflow-freespace" - this would limit some duplication and made it much more straightforward - how to run some commands locally. The parameters for those commands could still be passed by environment variables in CI (Click has support for that and we already use it).

The only problem to solve is how to handle errors when somene has "old"  "breeze" virtualenv. Breeze will change dynamically - new dependencies will be added and we will move to newer versions regularly. This might lead to unnecessary issues and  questions on Slack if those errorrs are somehow mysterious and not obvious. Current ./Breeze2 bootstrap script does it by automated upgrading of the virtualenv. However I thought about it and maybe we can approach it differently.

I think we have a way to handle it. We should write a little common code for those commands to capture and store in repo what dependencies are needed (and which versions - that could be even requirements.txt file) and update it automatically with pre-commit and warn the users in cases the dependencies are missing or outdated whenever any command is run. Those dependencies will not change often, but when they do, all users should know that they should upgrade and it should be possible.

This would be a warning only but it could provide a helpful "Your breeze enviroinment is outdated, please run `pipx install -e ./dev/breeze --force`  to update the environment" warning. 

Then we could (as @uranusjr proposed and stronly suggested) remove the custom `./Breeze2` bootstrapping. 

# Why it might be worse ?

There are two serious dangers with pipx:

* that the version you have in one branch will not work with another branch (very likely from the past experiences)
* that even if you are in the same branch, you could have not installed the latest version of dependencies (and Breeze will fail) or that you have not installed it with --editable flag - which means that the code of Breeze in the branch is different that what you run. This might lead to serious, unforeseen and difficult to diagnose problems.

Both are real dangers, and they will significantly increase the likelihood that different users (especially the ones who are not experienced) will get Breeze not working and that they will raise support questions or (worse) that they will consider Breeze as buggy and will avoid using it or even will drop the idea of contributing to Airflow.

The ./Breeze2 bootstrapping is free of those problems:

* --editable mode is automatically used
* latest dependencies are added when they were missing automatically
* your Breeze version is very tightly connecteect to the branch you are in

**Comment by Bowrna** [comment]
@potiuk I checked how airflow-freespace command is added as standalone commands.
Here the entry points in setup.cfg is :
```
airflow-freespace=airflow_ci.freespace:main
```
Do you want each command now available in breeze.py to go into a separate file with main function?

**Comment by potiuk** [comment]
Well. I thought a lot about it recently.

I am still not entirely "sold" on the idea which I described in the issue. 

This is especially afer comments of @nico in [This thread](https://apache-airflow.slack.com/archives/CQ9QHSFQX/p1642513096003800). 

He expressed some concerns which pretty much mirrored very similar to what I had initially when I wrote the https://github.com/apache/airflow/blob/main/dev/breeze/doc/adr/0003-bootstraping-virtual-environment.md ).

I thought I am the only one who thinks the "bootstrapping" is better (for those reasons) - but Niko explained pretty much all the reasons I put in the ADR why bootstrapping is better. 

Simply - while `pipx` is nice for installing stuff, it lacks in automated management and switching between branches. It will only install "one" installation and we really want to use the exact "breeze" version that comes with the actual sources (so for example when we checkout v2-2-test, we want to use Breeze that is installed there), when  we use (future) v2-3-test - we want to use Breeze version that is available in that branch. There are some "suffix" options but they are not easy to maintain and keep in check (and you have to remamber about using the right suffix in the right branch).

So I am currently torn in half (but leaning towards leaving the Bootstrapping as default approach):

1) `pipx` and multiple entrypoints are nice for installation and adding stuf on the PATH and are more "standard-ish" 
2) `./Breeze2` bootstrapping is nicer to make sure that the version uou are using is "latest version that works for this branch"  and does not need ANY extra tool

There are two serious dangers with `pipx`:

* that the version you have in one branch will not work with another branch (very likely from the past experiences)
* that even if you are in the same branch, you could have not installed the latest version of dependencies (and Breeze will fail) or that you have not installed it with `--editable` flag - which means that the code in the branch is different that what you installed. 

Both are real dangers, and they will significantly introduce the likelihood that different users (especially the ones who are not experienced) will get Breeze not working and that they will raise support questions or (worse) that they will consider Breeze as buggy and will avoid using it or even will drop the idea of contributing to Airflow.

The `./Breeze2` bootstrapping is free of those problems:

* --editable mode is automatically used
* latest dependencies are added when they were missing automatically 
* your Breeze version is very tightly connecteect to the branch you are in

Honestly, I think we can only go with the `pipx` solution if we have automated solutions to those problems. 

This could be done by some automated check at the start of each command: 

* whether we are installed in editable mode
* whether all dependencies are installed
* whether we are installed from the current branch 
* fixing it or "failing" with the right instruction to the user on how to fix

This is all possible to be implemented, but I have a feeiling that this will really outweight the benefits of using `pipx` in the first place. The complexity of such custom code is even bigger than the current `./Breeze2` bootstrapping script and somehow I have a feeling that the bootstrapping script is much more 'versatile' and "clener" because it is applied "before" and independently from Breeze code.

I would love to hear more comments on that but I have not made up my mind yet.

@uranusjr - WDYT?

Also maybe some other Breeze users @dstandish @ephraimbuddy @kaxil - WDYT?

**Comment by o-nikolas** [comment]
Thanks @potiuk (Niko here), you've captured my concerns very well :+1:
